opal/runtime/opal_progress.c

```diff
@@ -119,9 +151,33 @@ static void opal_progress_finalize(void)
     opal_atomic_unlock(&progress_lock);
 }
 
+void opal_progress_shutdown_async_progress_thread(void)
+{
+#if OPAL_ENABLE_PROGRESS_THREADS == 1
+    if (opal_async_progress_thread_spawned) {
+        /* shutdown the async thread */
+        thread_arg.running = false;
+        int err = opal_thread_join(&opal_progress_async_thread, NULL);
+        if (OPAL_SUCCESS == err) {
+            opal_set_using_threads(false);
+            opal_async_progress_thread_spawned = false;
+        } else {
+            OPAL_OUTPUT(
+                (debug_output, "progress: Failed to join async progress thread: err=%d", err));
+        }
+        // end of MPI process, reset nb_events_processed and
+        // not nb_events_reported - which was normally resetted by
+        // application threads
+        opal_atomic_swap_64(&thread_arg.nb_events_processed, 0);
```

Comment:
No, this is not necessarily the end of the MPI process. It is possible that opal_progress will continue to be called from the main thread, so it is important to let it know there were pending events it did not acknowledge yet.